packages/node/src/utils/kyve/kyve.ts

```diff
@@ -0,0 +1,530 @@
+
+      await timeout(
+        new Promise((resolve, reject) => {
+          writeStream.on('open', resolve);
+          writeStream.on('error', reject);
+        }),
+        BUNDLE_TIMEOUT,
+      );
+
+      const zippedBundleData = await this.retrieveBundleData(bundle.storage_id);
+
+      const gunzip = zlib.createUnzip({
+        maxOutputLength: MAX_COMPRESSION_BYTE_SIZE /* avoid zip bombs */,
+      });
+
+      logger.info(`Retrieving bundle ${bundle.id}`);
+
+      await new Promise((resolve, reject) => {
+        zippedBundleData.data
+          .pipe(gunzip)
+          .pipe(writeStream)
+          .on('error', reject)
+          .on('finish', async () => {
+            await fs.promises.chmod(bundleFilePath, 0o444);
+            resolve('Stream Completed');
+          });
+      });
+    } catch (e) {
+      if (!['EEXIST', 'EACCES'].includes(e.code)) {
+        await fs.promises.unlink(bundleFilePath);
+      }
+      throw e;
+    }
+  }
+
+  private async getBundleData(bundle: BundleDetails): Promise<string> {
+    const bundleFilePath = this.getBundleFilePath(bundle.id);
+    try {
+      await this.downloadAndProcessBundle(bundle);
+      return await this.readFromFile(bundleFilePath);
+    } catch (e: any) {
+      if (['EEXIST', 'EACCES'].includes(e.code)) {
+        const res = await this.pollUntilReadable(bundleFilePath);
+        return res;
+      } else {
+        throw e;
+      }
+    }
+  }
+
+  async readFromFile(bundleFilePath: string): Promise<string> {
+    return fs.promises.readFile(bundleFilePath, 'utf-8');
+  }
+
+  private isBundleFile(filename: string): boolean {
+    return BUNDLE_FILE_ID_REG(this.poolId).test(filename);
+  }
+
+  private async getExistingBundlesFromCacheDirectory(
+    tmpDir: string,
+  ): Promise<BundleDetails[]> {
+    const bundles: BundleDetails[] = [];
+    const files = await fs.promises.readdir(tmpDir);
+
+    for (const file of files) {
+      if (this.isBundleFile(file)) {
+        const id = parseDecimal(file.match(BUNDLE_FILE_ID_REG(this.poolId))[1]);
+        bundles.push(await this.getBundleById(id));
+      }
+    }
+
+    return bundles;
+  }
+
+  private async getToRemoveBundles(
+    cachedBundles: Record<string, Promise<BundleDetails>>,
+    height: number,
+    bufferSize: number,
+  ): Promise<BundleDetails[]> {
+    if (!Object.keys(cachedBundles).length) {
+      return this.getExistingBundlesFromCacheDirectory(this.tmpCacheDir);
+    }
+
+    const currentBundle = await this.getBundleFromCache(height);
+    if (!currentBundle) return [];
+
+    const bundles = await Promise.all(Object.values(cachedBundles));
+
+    return bundles.filter((b) => {
+      const isNotCurrentBundleAndLower =
+        currentBundle.id !== b.id &&
+        parseDecimal(currentBundle.id) > parseDecimal(b.id);
+      const isOutsiderBuffer =
+        height < parseDecimal(b.from_key) - bufferSize ||
+        height > parseDecimal(b.to_key) + bufferSize;
```

Comment:
I think only bundles below the buffer matter, If we're doing a binary search or have any future bundles they will be removed from the cache